webapp/channels/src/components/suggestion/provider.tsx

```diff
@@ -1,20 +1,33 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
-export type ProviderResult = {
+import {RequireOnlyOne} from '@mattermost/types/utilities';
+
+export type ProviderResult<Item> = {
     matchedPretext: string;
     terms: string[];
-    items: Array<Record<string, any>>;
-    component?: React.ReactNode;
-}
+    items: Array<Item | Loading>;
+} & RequireOnlyOne<{
+    component: React.ReactNode;
+    components: React.ReactNode[];
+}>;
+
+export type Loading = {
+    type: string;
+    loading: boolean;
+};
 
-export default class Provider {
+export type ResultsCallback<Item> = (result: ProviderResult<Item>) => void;
+
+export default abstract class Provider {
```

Comment:
Providers aren't React components, so we can still use inheritance for them. I have ideas for how we could possibly do this in a more React-friendly way in the future, but this PR already had enough scope creep as-is